c7n/resources/iam.py

```diff
@@ -37,10 +37,243 @@ class Role(QueryResourceManager):
 
 @resources.register('iam-user')
 class User(QueryResourceManager):
-
     resource_type = 'aws.iam.user'
 
 
+@resources.register('iam-policy')
+class Policy(QueryResourceManager):
+    resource_type = 'aws.iam.policy'
+
+
+@resources.register('iam-profile')
+class InstanceProfile(QueryResourceManager):
+    resource_type = 'aws.iam.instance-profile'
+
+
+@resources.register('iam-certificate')
+class ServerCertificate(QueryResourceManager):
+    resource_type = 'aws.iam.server-certificate'
+
+
+class IamRoleUsage(Filter):
+
+    def get_service_roles(self):
+        results = []
+        for result in self.scan_lambda_roles():
+            if result not in results:
+                results.append(result)
+        for result in self.scan_ecs_roles():
+            if result not in results:
+                results.append(result)
+        for result in self.scan_asg_roles(True):
+            if result not in results:
+                results.append(result)
+        for result in self.scan_ec2_roles(True):
+            if result not in results:
+                results.append(result)
+        return results
+
+    def get_instance_profiles(self):
+        results = []
+        for result in self.scan_asg_roles(False):
+            if result not in results:
+                results.append(result)
+        for result in self.scan_ec2_roles(False):
+            if result not in results:
+                results.append(result)
+        return results
+
+    def get_profile_role_arn(self, resource):
+        results = []
+        client = local_session(self.manager.session_factory).client('iam')
+        for iprofile in client.list_instance_profiles()['InstanceProfiles']:
+            if iprofile['Arn'] == resource:
+                profile = client.get_instance_profile(
+                    InstanceProfileName=iprofile['InstanceProfileName'])
+                if 'Roles' not in profile:
+                    return []
+                for role in profile['Roles']:
+                    results.append(role['Arn'])
+        return results
+
+    def scan_lambda_roles(self):
+        results = []
+        client = local_session(self.manager.session_factory).client('lambda')
+        fxs = client.list_functions()['Functions']
```

Comment:
you should be reusing the resource managers we already have, that ensure cache utilization.